src/index.js

```diff
@@ -81,32 +90,46 @@ class Dropzone extends React.Component {
       isDragReject: false
     });
 
-    const droppedFiles = e.dataTransfer ? e.dataTransfer.files : e.target.files;
-    const max = this.props.multiple ? droppedFiles.length : Math.min(droppedFiles.length, 1);
-    const files = [];
+    let droppedFiles = e.dataTransfer ? e.dataTransfer.files : e.target.files;
+    const dataTransferItems = e.dataTransfer && e.dataTransfer.items ? e.dataTransfer.items : [];
 
-    for (let i = 0; i < max; i++) {
-      const file = droppedFiles[i];
-      // We might want to disable the preview creation to support big files
-      if (!this.props.disablePreview) {
-        file.preview = window.URL.createObjectURL(file);
+    const wrapup = () => {
+      const max = this.props.multiple ? droppedFiles.length : Math.min(droppedFiles.length, 1);
+      const files = [];
+
+      for (let i = 0; i < max; i++) {
+        const file = droppedFiles[i];
+        // We might want to disable the preview creation to support big files
+        if (!this.props.disablePreview) {
+          file.preview = window.URL.createObjectURL(file);
+        }
+        files.push(file);
       }
-      files.push(file);
-    }
-
-    if (this.props.onDrop) {
-      this.props.onDrop.call(this, files, e);
-    }
 
-    if (this.allFilesAccepted(files)) {
-      if (this.props.onDropAccepted) {
-        this.props.onDropAccepted.call(this, files, e);
+      if (this.props.onDrop) {
+        this.props.onDrop.call(this, files, e);
       }
-    } else {
-      if (this.props.onDropRejected) {
-        this.props.onDropRejected.call(this, files, e);
+
+      if (this.allFilesAccepted(files)) {
+        if (this.props.onDropAccepted) {
+          this.props.onDropAccepted.call(this, files, e);
+        }
+      } else {
+        if (this.props.onDropRejected) {
+          this.props.onDropRejected.call(this, files, e);
+        }
       }
+    };
+
+    if (dataTransferItems[0] && typeof dataTransferItems[0].webkitGetAsEntry === 'function') {
+      const entry = dataTransferItems[0].webkitGetAsEntry();
+
+      return this.walkDirectory(entry.filesystem.root, walkedFiles => {
+        droppedFiles = walkedFiles;
+        wrapup();
+      });
     }
+    wrapup();
```

Comment:
This will be called 2 times, if I understand it correctly. 

Also, I don't like calling some function that returns nothing without arguments. This should be re-written to a something more testable and pure.

So, ultimately, this `wrapup` function should be a separate module (with its own tests) that takes files array as argument and returns files array with previews.